content/blog/2022-03-29-unified-gc/index.md

```diff
@@ -0,0 +1,115 @@
+
+## Reference Counting
+
+A reference counting collector keeps metadata that keeps track of how many references there are to each object. Reference counting runs every time a pointer is changed. When a pointer is assigned to point to an object, the count for that object is increased, and when it no longer points to an object the count is decreased. When the count goes to 0, the object can be freed. Low pause time, determinism, predictability, and simplicity are pros of reference counting.
+
+## Tracing
+
+Tracing does not keep track of any extra metadata. Tracing does not run every time a pointer is changed. Instead, it runs periodically as determined by the programming language. Although it does not run every time a pointer is changed and does not keep track of metadata, it must perform a graph traversal to identify which nodes are unreachable. Lower space overhead, control over when the collector should run, and cycle collection are pros of tracing. 
+
+# The Paper
+
+## Motivation
+
+Tracing and reference counting collectors each have their pros and cons. Researchers have explored the design space searching for optimizations. This paper presents a formalization of the design space, demonstrating that all garbage collection is a combination of reference counting and tracing. It also provides a way for collectors to perform analysis on the cost of such collectors.
+
+## Fix-Point Formulation
+
+Garbage collection is first defined using a fixed point formulation. Later in the paper, it is shown how tracing, reference counting, and hybrids of tracing and reference counting satisfy this fixed point formulation. We start out with a characterization of memory:
+* $V$ is the set of all objects. This includes objects that are still in use, objects that are no longer in use but not yet freed, and objects that have been freed. 
+* $E$ is the multiset of edges in the graph. In other words, the references between one object and another. It is a multi-set because an object can have multiple pointers to another node. Consider the case where object $a$ has fields $f1$ and $f2$ which both point to the same object $b$.
+* $R$ is the multiset of objects that are roots in the graph.
+* $p(v)$ where $v \in V$ is the reference count of vertex $v$.
+
+The object graph is the triple $G = <V, E, R>$. From here, garbage collection is given as a fixed point computation.
+
+<p align="center">
+<img src="FP.png" alt="alt_text" title="image_tooltip" style="zoom:25%;" />
+</p>
+
+Once reference counts have been assigned, vertices with counts of 0 are reclaimed. It is important to note that this fix point equation is not an algorithm. Instead, tracing, reference counting, and hybrids of the two are algorithms that satisfy this fix point equation.
```

Comment:
Again to be perhaps overly pedantic, they are algorithms _that find solutions_ that satisfy the equation.